mm2src/coins/qrc20/history.rs

```diff
@@ -0,0 +1,991 @@
+                ctx.log.log(
+                    "😟",
+                    &[&"tx_history", &self.utxo.ticker],
+                    &format!("'transfer' details of {:?} were reloaded", tx_hash),
+                );
+            }
+
+            mm_counter!(ctx.metrics, "tx.history.response.count", 1, "coin" => self.utxo.ticker.clone(), "method" => "transfer_details_by_hash");
+            if transactions_left > 0 {
+                transactions_left -= 1;
+                *unwrap!(self.utxo.history_sync_state.lock()) =
+                    HistorySyncState::InProgress(json!({ "transactions_left": transactions_left }));
+            }
+
+            updated = true;
+        }
+
+        *unwrap!(self.utxo.history_sync_state.lock()) = HistorySyncState::Finished;
+        updated
+    }
+
+    fn try_load_history_from_file(&self, ctx: &MmArc) -> HistoryMapByHash {
+        let history = self.load_history_from_file(&ctx);
+        let mut history_map: HistoryMapByHash = HashMap::default();
+
+        for tx in history {
+            let id = match TxInternalId::from_bytes(&tx.internal_id) {
+                Ok(i) => i,
+                Err(e) => {
+                    ctx.log.log(
+                        "😟",
+                        &[&"tx_history", &self.utxo.ticker],
+                        &ERRL!("Error {:?} on load history from file", e),
+                    );
+                    return HistoryMapByHash::default();
+                },
+            };
+            let tx_hash_history = history_map.entry(id.tx_hash.clone()).or_insert_with(HashMap::default);
+            if tx_hash_history.insert(id, tx).is_some() {
+                ctx.log.log(
+                    "😟",
+                    &[&"tx_history", &self.utxo.ticker],
+                    &ERRL!("History file contains entries with the same 'internal_id'"),
+                );
+                return HistoryMapByHash::default();
+            }
+        }
+
+        history_map
+    }
+}
+
+pub struct HistoryBuilder {
+    coin: Qrc20Coin,
+    from_block: i64,
+    topic: String,
+    address: H160,
+    token_address: H160,
+    order: Option<HistoryOrder>,
+}
+
+pub struct HistoryCont<T> {
+    history: Vec<T>,
+}
+
+/// Future for the [`HistoryBuilder::build_utxo_lazy()`].
+/// Loads `UtxoTx` from `tx_hash`.
+pub struct UtxoFromHashFuture {
+    future: Option<Box<dyn Future<Output = Result<UtxoTx, String>> + Unpin + 'static>>,
+}
+
+pub enum HistoryOrder {
+    NewestToOldest,
+    OldestToNewest,
+}
+
+impl HistoryBuilder {
+    pub fn new(coin: Qrc20Coin) -> HistoryBuilder {
+        let address = qrc20_addr_from_utxo_addr(coin.utxo.my_address.clone());
+        let token_address = coin.contract_address;
+        HistoryBuilder {
+            coin,
+            from_block: 0,
+            topic: QRC20_TRANSFER_TOPIC.to_string(),
+            address,
+            token_address,
+            order: None,
+        }
+    }
+
+    #[allow(clippy::wrong_self_convention)]
```

Comment:
Might worth to rename to `with_from_block` or `set_from_block` as `from_` fns are typically considered as constructors from other types.

Reply:
I agree that this attribute looks like a crutch, but it's common approach in `Builder pattern`.
The `HistoryBuilder` has the same method as [FilterBuilder](https://github.com/tomusdrw/rust-web3/blob/master/src/types/log.rs#L104). I could rename this to `search_from_block`, but it will be confusing.